### DISCUSSION TITLE
Which PIO StateMachines are used by MicroPython for the Pico W?

### DISCUSSION THREAD

**Discussion by ZacharyInDallas** [micropython/micropython#13044] [RP2040 / Pico]
I need to pick PIO StateMachine numbers.  I've seen a post that suggested that WiFi uses PIO StateMachine 1, but I'm not certain about that.

Is there any documentation about which StateMachines are used by MicroPython itself?

Is there a way to tell programatically which state machines are in use?

I tried doing an .active () on each StateMachine.  That threw an exception of 2 StateMachines but I don't know how to interpret that.

I tried to search the MicroPython source and could not find any search keys that turned up anything.  What should be looked for?

**Comment by jimmo** [comment]
The wifi driver uses PIO 1, but will fall back to PIO 0. See https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_cyw43_driver/cyw43_bus_pio_spi.c#L72

From looking at the code it, it appears to only claims a single SM from the chosen PIO.

Nothing in MicroPython uses PIO by default.

**Comment by dhalbert** [comment]
We were recently wondering why the CYW43 drivers uses PIO SPI instead of hardware SPI. We are running out of PIO to also do USB Host. The PIO SPI could be due to pin restrictions, to wanting to leave both SPI0 and SPI1 free, or to something special about the SPI capability needed. Do you happen to know? Thanks.

**Reply by jimmo** [reply]
>  something special about the SPI capability needed

@dhalbert yep exactly that -- it's "single wire" spi (i.e. SDI and SDO are the same pin, GPIO24). It's a lot of effort to save a single pin.
